apps/backend/src/purchases/purchase.processor.ts

```diff
@@ -0,0 +1,80 @@
+import { Process, Processor, OnQueueActive } from '@nestjs/bull';
+import { Logger } from '@nestjs/common';
+import { Purchase } from '@prisma/client';
+import { Job } from 'bull';
+import polly = require('polly-js');
+
+import { IssuerService } from '../issuer/issuer.service';
+import { PrismaService } from '../prisma/prisma.service';
+import { PurchasesService } from './purchases.service';
+import { getClaimData } from './utils';
+
+const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+export interface ClaimPurchaseData {
+  purchaseId: Purchase['id'];
+  previousTx?: string; // Previous transaction that needs to be waited on
+}
+
+@Processor('purchase')
+export class PurchaseProcessor {
+  private readonly logger = new Logger(PurchaseProcessor.name, {
+    timestamp: true,
+  });
+
+  constructor(
+    private prisma: PrismaService,
+    private readonly purchasesService: PurchasesService,
+    private readonly issuerService: IssuerService,
+  ) {}
+
+  @OnQueueActive()
+  onQueueActive(jobId: number) {
+    this.logger.debug(`Received job ${JSON.stringify(jobId)}`);
+  }
+
+  @Process('claim')
+  async claimPurchase({ data: { purchaseId, previousTx } }: Job<ClaimPurchaseData>) {
+    const purchase = await this.purchasesService.findOne(purchaseId);
+
+    if (previousTx) {
+      await this.waitForTxMined(previousTx);
+    }
+
+    const txHash = await this.issuerService.claimCertificate({
+      id: Number(purchase.certificate.onchainId),
```

Comment:
why don't we store `purchase.certificate.onchainId` as a number ?

Reply:
Its a bigger discussion, but thats because JS number supports up to 2^53 and solidity 2^256. I can explain more on the daily if needed.